log-manager/src/main/java/io/airlift/log/RollingFileMessageOutput.java

```diff
@@ -444,4 +459,44 @@ else if (fileName.startsWith(DELETED_PREFIX)) {
         }
         return files.build();
     }
+
+    @VisibleForTesting
+    static class FileTimeRoller
+    {
+        public static final FileTimeRoller SYSTEM_FILE_TIME_ROLLER = new FileTimeRoller(System::currentTimeMillis, Ticker.systemTicker());
+
+        private final LongSupplier systemTimeMillis;
+        private final Ticker ticker;
+
+        @GuardedBy("this")
+        private long nextRollNanos;
+
+        public FileTimeRoller(LongSupplier systemTimeMillis, Ticker ticker)
+        {
+            this.systemTimeMillis = requireNonNull(systemTimeMillis, "systemTimeMillis is null");
+            this.ticker = requireNonNull(ticker, "ticker is null");
+        }
+
+        public LocalDateTime currentDateTime()
+        {
+            return LocalDateTime.ofInstant(Instant.ofEpochMilli(systemTimeMillis.getAsLong()), ZoneId.systemDefault());
+        }
+
+        public synchronized boolean shouldRoll()
+        {
+            return ticker.read() >= nextRollNanos;
+        }
+
+        @VisibleForTesting
+        synchronized long getNextRollNanos()
+        {
+            return nextRollNanos;
+        }
+
+        public synchronized void updateNextRollTime(LocalDateTime logFileCreateTime)
+        {
+            long epochMilli = logFileCreateTime.toLocalDate().plusDays(1).atStartOfDay(ZoneId.systemDefault()).toInstant().toEpochMilli();
+            nextRollNanos = ticker.read() + TimeUnit.MILLISECONDS.toNanos(epochMilli - systemTimeMillis.getAsLong());
```

Comment:
This can overflow, and subsequent comparison with `nextRollNanos` will be broken. Specifically, from the System.nanoTime() javadocs:


> To compare elapsed time against a timeout, use
>  `if (System.nanoTime() - startTime >= timeoutNanos) ...`
> instead of
>  `if (System.nanoTime() >= startTime + timeoutNanos) ...`
> because of the possibility of numerical overflow.
```